prov/gni/src/gnix_nic.c

```diff
@@ -52,12 +52,231 @@ static pthread_mutex_t gnix_nic_list_lock = PTHREAD_MUTEX_INITIALIZER;
 uint32_t gnix_def_max_nics_per_ptag = 4;
 
 /*
- * stub for now
+ * function to process GNI CQ TX CQES to progress a gnix_nic
  */
 
+static int __nic_tx_progress(struct gnix_nic *nic)
+{
+	int ret = FI_SUCCESS;
+	int msg_id=0;
+	gni_return_t  status = GNI_RC_NOT_DONE, status2;
+	gni_post_descriptor_t  *gni_desc=NULL;
+	gni_cq_entry_t cqe;
+	struct gnix_tx_descriptor *gnix_tdesc = NULL;
+	unsigned int recov;
+
+	/*
+	 * first see if there are any CQE's on the
+	 * tx cq. Calling GNI_CqTestEvent does not require
+	 * getting the nic lock
+	 */
+
+	status = GNI_CqTestEvent(nic->tx_cq);
+	if (status == GNI_RC_NOT_DONE)
+		return FI_SUCCESS;
+
+try_again:
+	fastlock_acquire(&nic->lock);
+        status = GNI_CqGetEvent(nic->tx_cq, &cqe);
+        if (status  == GNI_RC_NOT_DONE) {
+		fastlock_release(&nic->lock);
+		return FI_SUCCESS;
+	}
+
+	switch (status) {
+	case GNI_RC_SUCCESS:
+		assert(GNI_CQ_STATUS_OK(cqe));
+		/*
+		 * check whether CQE from SMSG or a Post
+		 * transaction
+		 */
+		if (GNI_CQ_GET_TYPE(cqe)
+				== GNI_CQ_EVENT_TYPE_POST) {
+			status2 = GNI_GetCompleted(nic->tx_cq, cqe, &gni_desc);
+			if ((status2 != GNI_RC_SUCCESS) &&
+				(status2 != GNI_RC_TRANSACTION_ERROR)) {
+				ret = gnixu_to_fi_errno(status2);
+			}
+			gnix_tdesc = container_of(gni_desc,
+						struct gnix_tx_descriptor,
+						desc.gni_desc);
+		}  else if (GNI_CQ_GET_TYPE(cqe)
+			    == GNI_CQ_EVENT_TYPE_SMSG) {
+			msg_id = GNI_CQ_GET_MSG_ID(cqe);
+			gnix_tdesc = gnix_desc_lkup_by_id(nic,msg_id);
+			if (gnix_tdesc == NULL)
+				ret = -FI_ENOENT;
```

Comment:
This return path leaves the function while still holding the lock.  I don't think it's intentional, but if it is, I would strongly suggest a redesign.